VE 1920: Port Source Mode Editor Feature

pauloslund opened the pull request.

Commit:
VE-1920: Get SourceModeDialog to show

Commit:
VE-1920: Tear down old VE and replace reinitialize

Commit:
VE-1920: Configure proper i18n messages for SourceMode Dialog

Commit:
VE-1920: Cleanup JSHint error

inez commented on extensions/VisualEditor/lib/ve/src/ui/ve.ui.Surface.js:
```diff
@@ -32,6 +32,8 @@ ve.ui.Surface = function VeUiSurface( dataOrDoc, config ) {
 	// Mixin constructor
 	OO.EventEmitter.call( this, config );
 
+	this.target = target;
```
Should be under comment: // Properties

inez commented on extensions/VisualEditor/lib/ve/src/ui/ve.ui.Surface.js:
```diff
@@ -21,7 +21,7 @@
  * @cfg {string} [placeholder] Placeholder text to display when the surface is empty
  * @cfg {string} [inDialog] The name of the dialog this surface is in
```
Missing comment about new param

inez commented on extensions/VisualEditor/modules/ve-mw/init/ve.init.mw.Target.js:
```diff
@@ -1402,6 +1406,59 @@ ve.init.mw.Target.prototype.tryWithPreparedCacheKey = function ( doc, options, e
+ * @param {Object} [settings] Additional AJAX settings, or overrides of default settings
+ * @return {jqXHR} Return value of $.ajax()
+ */
+ve.init.mw.Target.static.apiRequest = function ( data, settings ) {
```
Later we can move it to our target just to minimize amount of change to the core

inez commented on extensions/VisualEditor/modules/ve-mw/init/ve.init.mw.Target.js:
```diff
@@ -1893,3 +1950,62 @@ ve.init.mw.Target.prototype.scrollToHeading = function ( headingNode ) {
+ * @method
+ * @fires sanityCheckComplete
+ */
+ve.init.mw.Target.prototype.startSanityCheck = function () {
```
Why there was a need to port this method? There is no sanity check anymore in new VE? If there is no then maybe worth looking into their git to find out why the resigned form it - and we should do it too.

inez commented on extensions/VisualEditor/wikia/modules/ve/ui/dialogs/ve.ui.WikiaSourceModeDialog.js:
```diff
@@ -0,0 +1,216 @@
+		mw.loader.using(
+			'ext.wikia.LinkSuggest',
+			function () {
+				this.sourceModeTextarea.$input.linksuggest();
```
I don't know why it was like this, but perhaps just this.sourceModeTextarea.$input.linksuggest.bind(this) should work

pauloslund replied:
Was done like this because `this.sourceModeTextarea.$input.linksuggest` does not exist until after `ext.wikia.LinkSuggest` has loaded so it cannot be bound to.

inez commented on extensions/VisualEditor/wikia/modules/ve/ui/dialogs/ve.ui.WikiaSourceModeDialog.js:
```diff
@@ -0,0 +1,216 @@
+	this.sourceModeTextarea.$input[0].setSelectionRange( 0, 0 );
+	this.$content.stopThrobbing();
+
+  // ve-upstream-sync - review - @author: Paul Oslund
```
Something off with the indentation here

inez commented on extensions/VisualEditor/wikia/modules/ve/ui/tools/ve.ui.WikiaDialogTool.js:
```diff
@@ -0,0 +1,25 @@
+ve.ui.WikiaSourceModeDialogTool.static.name = 'wikiaSourceMode';
+ve.ui.WikiaSourceModeDialogTool.static.group = 'object';
+ve.ui.WikiaSourceModeDialogTool.static.icon = 'source';
+ve.ui.WikiaSourceModeDialogTool.static.title = 'Source Mode';
```
Is there something stopping you from start using OO.ui.deferMsg? If so then please put review comment here so we don't end up going on production with hardcoded string.

inez commented on extensions/VisualEditor/wikia/modules/ve/ui/dialogs/ve.ui.WikiaSourceModeDialog.js:
```diff
@@ -0,0 +1,216 @@
+		target.toolbarSaveButton.disconnect( target );
+		target.toolbarSaveButton.$element.detach();
+		target.getToolbar().$actions.empty();
+		target.teardownSurface( true ).done( function () {
```
Just a reminder about that not needed true

inez replied:
Just a reminder about that not needed true

pauloslund replied:
This was changed so that true is needed again. I added in the noAnimate flag so that we weren't relying on the lack of CSS transitions to prevent an animation

Commit:
VE-1920: Cleanup code structure and add i18n for SourceMode tooltip

Commit:
VE-1920: Re-add noAnimate flag to future-proof CSS changes

Commit:
VE-1920: Remove unnecessary sanityCheck

This was initially remove here: https://github.com/wikimedia/mediawiki-extensions-VisualEditor/commit/237bb05 because it was slow.

inez commented on extensions/VisualEditor/modules/ve-mw/init/targets/ve.init.mw.DesktopArticleTarget.js:
```diff
@@ -961,20 +962,28 @@ ve.init.mw.DesktopArticleTarget.prototype.restoreScrollPosition = function () {
-			deferred.resolve();
+		deferred = $.Deferred(),
+		tearDownToolbar = function () {
+		target.toolbar.destroy();
```
Something off with the indentations. Also, when it comes to more complex constructs, like functions, we tend to intialize them separately - instead of doing it with all the other vars - seems to make code a bit more readable.

pauloslund replied:
I agree with breaking functions into their own `var` statement; unfortunately the JSCS grunt task fails when it is initialized separately.

Commit:
VE-1920: Indentation fixes

dianafa commented on extensions/VisualEditor/wikia/modules/ve/ui/styles/ve.ui.WikiaSourceModeDialog.css:
```diff
@@ -0,0 +1,17 @@
+}
+.ve-ui-wikiaSourceModeDialog-content .oo-ui-window-body .oo-ui-textInputWidget .ve-ui-wikiaSourceModeDialog-helplink {
+	font-weight: bold;
+	line-height: 3em;
```
Do we have any styleguide determining font-sizes, paddings etc?
Also, is this intended to mix `em` and `px` units?
Why we decided to use pure css instead of sass?

pauloslund replied:
As it stands, we cannot currently load sass files; but once we can this file will be replaced with sass. Font-sizes, paddings, etc will then be variable driven.

dianafa commented on extensions/VisualEditor/wikia/modules/ve/ui/dialogs/ve.ui.WikiaSourceModeDialog.js:
```diff
@@ -0,0 +1,215 @@
+		dataType: 'json',
+		type: 'POST',
+		// Wait up to 100 seconds before giving up
+		timeout: 100000,
```
wow, was it always 100sec?

pauloslund replied:
Haha, unfortunately yes...

dianafa commented on extensions/VisualEditor/wikia/modules/ve/ui/dialogs/ve.ui.WikiaSourceModeDialog.js:
```diff
@@ -0,0 +1,215 @@
+		// jquery.ui.autocomplete.js
+		this.sourceModeTextarea.$input.css( {
+			'z-index': 9999999,
+			position: 'relative'
```
we can add `'` here for consistency :)

dianafa commented on extensions/VisualEditor/modules/ve-mw/init/ve.init.mw.Target.js:
```diff
@@ -1348,10 +1352,10 @@ ve.init.mw.Target.prototype.tryWithPreparedCacheKey = function ( doc, options, e
 			.then(
-				function ( response, jqxhr ) {
+				function ( response, status, jqxhr ) {
 					var eventData = {
```
Are we going to need to use this status?

pauloslund replied:
We don't need need the status; however, the parameters passed into this function have changed to include the status so we need to represent it here.

Commit:
VE-1920: Fix issue with incosistent keys
